src/deployer.ts

```diff
@@ -0,0 +1,194 @@
+    )
+    .describe('write-agent-config', 'Whether or not to write agent config')
+    .default('write-agent-config', true)
+    .boolean('write-agent-config').argv;
+}
+
+type MultisigIsmContracts = {
+  multisigIsm: LegacyMultisigIsm;
+};
+
+export class HyperlanePermissionlessDeployer {
+  constructor(
+    protected multiProvider: MultiProvider,
+    protected signer: ethers.Signer,
+    protected local: ChainName,
+    protected remotes: ChainName[],
+    protected writeAgentConfig?: boolean,
+  ) {}
+
+  static async fromArgs(): Promise<HyperlanePermissionlessDeployer> {
+    const multiProvider = getMultiProvider();
+    const { local, remotes, key, writeAgentConfig } = await getArgs(
+      multiProvider,
+    );
+    const signer = new ethers.Wallet(key);
+    multiProvider.setSharedSigner(signer);
+
+    return new HyperlanePermissionlessDeployer(
+      multiProvider,
+      signer,
+      local,
+      remotes as unknown as string[],
+      writeAgentConfig,
+    );
+  }
+
+  get chains(): ChainName[] {
+    return this.remotes.concat([this.local]);
+  }
+
+  async deploy(): Promise<void> {
+    let contracts: ChainMap<HyperlaneContracts> = {};
+    const owner = await this.signer.getAddress();
+    // First, deploy core contracts to the local chain
+    // NB: We create core configs for *all* chains because
+    // we also use coreDeployer to deploy MultisigIsms.
+    // Once we move that out to a HyperlaneIsmDeployer
+    // we can just do:
+    // const coreContracts = await coreDeployer.deploy();
+    const coreConfig = buildCoreConfig(owner, this.chains);
+    const coreDeployer = new HyperlaneCoreDeployer(
+      this.multiProvider,
+      coreConfig,
+    );
+    const coreContracts: ChainMap<CoreContracts> = {};
+    coreContracts[this.local] = await coreDeployer.deployContracts(
+      this.local,
+      coreConfig[this.local],
+    );
+    contracts = objMerge(contracts, coreContracts);
+
+    // Next, deploy MultisigIsms to the remote chains
+    // TODO: Would be cleaner if using HyperlaneIsmDeployer
+    const isms: ChainMap<MultisigIsmContracts> = {};
+    isms[this.local] = {
+      multisigIsm: coreContracts[this.local].multisigIsm,
+    };
+    for (const remote of this.remotes) {
+      isms[remote] = {
+        multisigIsm: await coreDeployer.deployLegacyMultisigIsm(remote),
+      };
+    }
+    contracts = objMerge(contracts, isms);
+
+    // Next, deploy TestRecipients to all chains
+    const testRecipientConfig: ChainMap<TestRecipientConfig> = objMap(
+      isms,
+      (chain, ism) => {
+        return { ism: ism.multisigIsm.address };
+      },
+    );
+    const testRecipientDeployer = new HyperlaneTestRecipientDeployer(
+      this.multiProvider,
+      testRecipientConfig,
+    );
+    const testRecipients = await testRecipientDeployer.deploy();
+    contracts = objMerge(contracts, testRecipients);
+
+    // Finally, deploy IGPs to all chains
```

Comment:
why do we need IGPs on all chains? fragmenting IGP addresses might get ugly

Reply:
Unfortunately we need it (for now) due to the 1:1 relationship between IGPs and relayers

If we modified the IGP payForGas interface to include a beneficiary then everyone could use the same contract.